radial_build_tools.py

```diff
@@ -0,0 +1,474 @@
+import matplotlib.pyplot as plt
+from matplotlib.patches import Rectangle
+import matplotlib.colors
+import yaml
+import argparse
+import numpy as np
+import openmc
+import textwrap
+
+
+class RadialBuildPlot(object):
+    """
+    Uses a radial build definition to generate radial build plots.
+
+    Parameters
+        build (dict): {"layer name": {
+                            "thickness": (float),
+                            "composition": {
+                                "material name": fraction (float)
+                                },
+                            "description": (str),
+                    }
+                }
+            The dict corresponding to each "layer_name" key may be empty,
+            or have any combination of entries.
+        title (string): title for plot and filename to save to
+        colors (list of str): list of matplotlib color strings.
+            If specific colors are desired for each layer they can be added
+            here.
+        max_characters (float): maximum length of a line before wrapping the
+            text
+        max_thickness (float): maximum thickness of layer to display, useful
+            for reducing the total size of the figure.
+        size (iter of float): figure size, inches. (width, height)
+        unit (str): Unit of thickness values
+    """
+
+    def __init__(
+        self,
+        build,
+        title="radial_build",
+        colors=None,
+        max_characters=35,
+        max_thickness=1e6,
+        size=(8, 4),
+        unit="cm",
+    ):
+        self.build = build
+        self.title = title
+        if colors is None:
+            self.colors = list(matplotlib.colors.XKCD_COLORS.values())[
+                0 : len(self.build)
+            ]
+        self.max_characters = max_characters
+        self.max_thickness = max_thickness
+        self.size = size
+        self.unit = unit
+
+    def build_composition_string(self, composition):
+        """
+        Assembles string from composition dict for use in radial build plot
+
+        Arguments:
+            composition (dict): "material name (str)":volume_fraction (float)
+
+        Returns:
+            comp_string (string): formatted string with composition definition
+        """
+
+        """comp_string = ""
+        for material, fraction in composition.items():
+
+            mat_string = f"{material}: {round(fraction*100, 3)}%, "
+            comp_string += mat_string
+        comp_string = textwrap.fill(
+            comp_string, width=self.max_characters, drop_whitespace=False
+        )
+
+        return comp_string[0:-2] + "\n
+        """
+        mat_strings = [
+            f"{mat}: {round(frac*100,3)}%" for mat, frac in composition.items()
+        ]
+
+        comp_string = (
+            textwrap.fill(", ".join(mat_strings), width=self.max_characters)
+            + "\n"
```

Comment:
In the case where composition is not specified, I don't want a newline in the text where the composition would go if it were specified.

`text = f"{name}{thickness_str}\n{comp_string}{description_str}".rstrip()`

Currently, it is guaranteed that there will be at the very least a name for each layer, the rest being optional. I add the one newline after the name and thickness strings as those two always go on the same line. If there are no additional elements the newline gets stripped, otherwise they end up on subsequent lines.

I add the newline in `build_composition_string()`  rather than when putting the strings together so it only gets added when there is a composition specified. If there is no composition, but there is a description, adding the newline when joining the strings will result in 2 newlines between the name and the description string.